design/one-pager-fn-claim-conditions.md

```diff
@@ -0,0 +1,255 @@
+                        "authorized to use it.",
+}
+desiredXR.Resource.SetConditions(c)
+response.SetDesiredCompositeResource(rsp, desiredXR)
+```
+
+## Proposal
+We would like to allow the Composition Function author to:
+- Choose where results go (Claim or XR)
+- Allow results to update the Status Conditions of the XR and Claim
+
+The following sections get into the details of each of the above items.
+
+### Choose Where Results Go
+Currently each result returned by a function will create a corresponding
+event on the XR (if no previous fatal result exist).
+
+We can expand this functionality by allowing the Result to have targets. In
+order to accomplish this, we will need to expand the Result API as follows.
+```protobuf
+message Result {
+  // Omitted for brevity
+
+  Target target = 3;
+}
+
+// Target of Function results.
+enum Target {
+  TARGET_UNSPECIFIED = 0;
+  TARGET_COMPOSITE_ONLY = 1;
+  TARGET_COMPOSITE_AND_CLAIM = 2;
+}
+```
+The reason for having `TARGET_COMPOSITE_AND_CLAIM` and not `TARGET_CLAIM` is an
+implementation limitation. This prevents more involved API changes, and this
+is also consistent with existing behavior (func copies to XR, Crossplane copies
+XR to Claim).
+
+An example of a function using this behavior to create events for both the
+Composite and Claim:
+```go
+// rb "github.com/crossplane/function-sdk-go/response/result/builder"
+// var messageUnauthorized = errors.New("You are unauthorized to access this resource.")
+result := rb.Fatal(messageUnauthorized).
+  TargetCompositeAndClaim().
+  Build()
+response.AddResult(rsp, result)
+```
+
+### Allow Results to Set a Condition
+We would like the function author to be able to set the Claim's status
+conditions. This would allow the function author to clearly communicate the
+state of the Claim with their users.
+
+To allow the setting of conditions in the result, we will need to expand the
+Result API as follows.
+```protobuf
+message Result {
+  // Omitted for brevity
+
+  // Optionally update the supplied status condition on all targets.
+  // The result's reason and message will be used in the condition.
+  optional Condition condition = 4;
+}
+
+message Condition {
+  // Type of the condition, e.g. DatabaseReady.
+  // 'Ready' and 'Synced' are reserved for use by Crossplane.
+  string type = 1;
+
+  // Status of the condition.
+  Status status = 2;
+
+  // Machine-readable PascalCase reason.
+  string reason = 3;
+}
+```
+
+An example of a function utilizing this new ability:
+```go
+// rb "github.com/crossplane/function-sdk-go/response/result/builder"
+// const databaseReady = "DatabaseReady"
+// const reasonUnauthorized = "Unauthorized"
+// var messageUnauthorized = errors.New("You are unauthorized to access this resource.")
+result := rb.Fatal(messageUnauthorized).
+  TargetCompositeAndClaim().
+  WithConditionFalse(databaseReady, reasonUnauthorized).
+  Build()
+response.AddResult(rsp, result)
+```
```

Comment:
FWIW I'm not sure I want to adopt the builder pattern in the Go SDK. That said, this is fine as an illustration. 🙂

Reply:
> FWIW I'm not sure I want to adopt the builder pattern in the Go SDK

I'm not sold on it either; I'm open to ideas.
The idea behind the builder was due to the Result being more complex at this point and the handling of args was beginning to get a little messy IMO.

Here are some thoughts on the subject:
```go
// expose all args. Not great as condition and target (should?) likely be optional
response.Normal(rsp, message, target, conditionType, conditionStatus, conditionReason)

// we could make both target and condition pointers, handling for nil
condition := Condition(type, status, reason)
response.Normal(rsp, message, target, condition)

// Perhaps an opts approach?
response.Normal(rsp, message, ...opts)

// We could supply predefined opts like Condition(type, status, reason) and Target_X_()
response.Normal(rsp, message, TargetClaim(), Condition("DatabaseReady", false, "Unauthorized"))
```

Reply:
I like the variadic functions as arguments approach you've got sketched out in the last alternative best. The only gotcha would be `response.Normalf` etc, which already accept variadic format args.

I think there's some precedent for a slightly lighter weight version of the builder pattern in crossplane runtime, e.g. we have:

```go
c := xpv1.Available().WithMessage("foo")
```

Where `WithMessage` returns the underlying condition with the message set.

So I could probably live with:

```go
response.Normalf(rsp, message, args...).
    TargetClaim().
    Condition("DatabaseReady, false, "Unauthorized")
```

Reply:
@dalton-hill-0 Could you update the design doc to make it clear that the builder pattern example is just a sketch, not necessarily the API we're committing to.